src/lib/notesManager.ts

```diff
@@ -0,0 +1,227 @@
+  binaryId: string;
+}
+
+export class NotesManager {
+  private viamClient: VIAM.ViamClient;
+  private machineId: string;
+
+  constructor(viamClient: VIAM.ViamClient, machineId: string) {
+    this.viamClient = viamClient;
+    this.machineId = machineId;
+  }
+
+  /**
+   * Save a note for a specific pass
+   */
+  async savePassNote(passId: string, noteText: string, partId: string): Promise<void> {
+    if (!this.viamClient) {
+      throw new Error("Viam client not initialized");
+    }
+
+    const now = new Date();
+    console.log(`Saving note for pass ${passId}: "${noteText}"`);
+
+    if (!partId) {
+      throw new Error("No part ID available for upload");
+    }
+
+    const noteData: PassNote = {
+      pass_id: passId,
+      note_text: noteText,
+      created_at: now.toISOString(),
+      created_by: "web-app"
+    };
+
+    const noteJson = JSON.stringify(noteData);
+    const binaryData = new TextEncoder().encode(noteJson);
+
+    await this.viamClient.dataClient.binaryDataCaptureUpload(
+      binaryData,
+      partId,
+      "rdk:component:generic",
+      "sanding-notes",
+      "SaveNote",
+      ".json",
+      [now, now],
+      ["sanding-notes"]
+    );
+
+    console.log("Note saved successfully!");
+
+    // After saving, clean up old notes for the same pass
+    await this.deleteOldNotesForPass(passId);
+  }
+
+  /**
+   * Deletes all but the most recent note for a given pass.
+   */
+  async deleteOldNotesForPass(passId: string): Promise<void> {
+    if (!this.viamClient) {
+      throw new Error("Viam client not initialized");
+    }
+
+    console.log(`Cleaning up old notes for pass ${passId}`);
+
+    const filter = {
+      robotId: this.machineId,
+      componentName: "sanding-notes",
+      componentType: "rdk:component:generic",
+      tags: ["sanding-notes"],
+    } as unknown as VIAM.dataApi.Filter;
+
+    const allNotesForPass: PassNoteWithId[] = [];
+    let paginationToken: string | undefined = undefined;
+    let hasMore = true;
+
+    // Fetch all notes metadata
+    while (hasMore) {
+      const binaryData = await this.viamClient.dataClient.binaryDataByFilter(
+        filter,
+        100,
+        VIAM.dataApi.Order.DESCENDING,
+        paginationToken,
+        false
+      );
+
+      const promises = binaryData.data.map(async (item) => {
+        try {
+          const noteDataArray = await this.viamClient.dataClient.binaryDataByIds([item.metadata!.binaryDataId!]);
+          if (noteDataArray.length > 0 && noteDataArray[0].binary) {
+            const noteJson = new TextDecoder().decode(noteDataArray[0].binary);
+            const noteData = JSON.parse(noteJson) as PassNote;
+            if (noteData.pass_id === passId) {
+              allNotesForPass.push({ ...noteData, binaryId: item.metadata!.binaryDataId! });
+            }
+          }
+        } catch (parseError) {
+          console.warn("Failed to parse note data during cleanup:", parseError);
```

Comment:
If we are going to do something like this we should also validate that it has the `tags: ["sanding-notes"]` tags